### ISSUE TITLE
[golang client sdk v3.5.0] Client v3.5.0 may be not exactly thread safe during retryAuth

### ISSUE THREAD

**Issue by chenaoxd** [etcd-io/etcd#13305]
### Description
Sometimes, the client will give `etcdserver: user name is empty` error. After some investigation, I think this may relate to the `unaryClientIterceptor`, and describe as below.

After the token expires, the `unaryClientIterceptor` will try to reauth the client and update the token. When updating the client will first [set the auth token to empty string](https://github.com/etcd-io/etcd/blob/ef81030105b5519b49de6d13218c32532d932db6/client/v3/retry_interceptor.go#L82), requests during this period may return `etcdserver: user name is empty` ?

### Versions
v3.5.0, `go.mod` as below
```
   go.etcd.io/etcd/api/v3 v3.5.0
   go.etcd.io/etcd/client/v3 v3.5.0
```

### Code to reproduce this issue
```go
package main

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"go.etcd.io/etcd/client/v3"
)

func main() {
	etcdConfig := clientv3.Config{
		Endpoints:   []string{"somehost"},
		DialTimeout: 10 * time.Second,
		Username:    "someuser",
		Password:    "somepass",
	}
	client, err := clientv3.New(etcdConfig)
	if err != nil {
		fmt.Println(err)
		return
	}

	for {
		for i := 0; i < 30; i++ {
			go func() {
				resp, err := client.Get(context.Background(), "/test")
				if err != nil {
					fmt.Println(err)
					return
				}
				fmt.Printf("res: %d\n", len(resp.Kvs))
				for _, kv := range resp.Kvs {
					fmt.Printf("%s -> %s\n", "/test", kv.Value)
				}
			}()
			time.Sleep(time.Duration(rand.Float32()*10000) * time.Microsecond)
		}
		time.Sleep(6 * time.Minute)
	}
}

```

### Logs (After 5 minutes sleeping)

```txt
{"level":"warn","ts":"2021-08-26T19:20:28.811+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.812+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.812+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.812+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.812+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.812+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.812+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.812+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.871+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.871+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
{"level":"warn","ts":"2021-08-26T19:20:28.911+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = InvalidArgument desc = etcdserver: user name is empty"}
etcdserver: user name is empty
{"level":"warn","ts":"2021-08-26T19:20:28.911+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = InvalidArgument desc = etcdserver: user name is empty"}
{"level":"warn","ts":"2021-08-26T19:20:28.911+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001f7340/#initially=[gpu.powerlaw.ai:42379]","attempt":0,"error":"rpc error: code = InvalidArgument desc = etcdserver: user name is empty"}
etcdserver: user name is empty
etcdserver: user name is empty
```

**Comment by cfc4n** [comment]
Thanks for report. I'll test it later.

**Comment by cfc4n** [comment]
It‘s not a bug.

[The token has expired or is invalid will be emptied](https://github.com/etcd-io/etcd/blob/ef81030105b5519b49de6d13218c32532d932db6/client/v3/retry_interceptor.go#L82), and you need to login again.

When the client has not completed login and authentication, Cannot be directly used for query.


Your problem is that run 30 coroutines, and they did not wait for the client finish logging in.

Try it.

```go
package main

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"go.etcd.io/etcd/client/v3"
)

func main() {
	etcdConfig := clientv3.Config{
		Endpoints:   []string{"http://127.0.0.1:2379"},
		DialTimeout: 10 * time.Second,
		Username:    "root",
		Password:    "root",
	}
	client, err := clientv3.New(etcdConfig)
	if err != nil {
		fmt.Println(err)
		return
	}

	for {
		for i := 0; i < 30; i++ {
			//go func() {
			//}()
			resp, err := client.Get(context.Background(), "/test")
			if err != nil {
				fmt.Println(err)
				return
			}
			fmt.Printf("res: %d\n", len(resp.Kvs))
			for _, kv := range resp.Kvs {
				fmt.Printf("%s -> %s\n", "/test", kv.Value)
			}
			time.Sleep(time.Duration(rand.Float32()*10000) * time.Microsecond)
		}
		// ./etcd --auth-token-ttl=30
		time.Sleep(40 * time.Second)
	}
}
```

**Comment by chenaoXD** [comment]
But the client will login automatically, and I don't think the user should do anything manually to wait for the login completed.

The document of the old version (like v3.3.24+incompatible) etcd golang SDK says that `The Client has internal state (watchers and leases), so Clients should be reused instead of created as needed. Clients are safe for concurrent use by multiple goroutines.` ([link](https://pkg.go.dev/go.etcd.io/etcd/clientv3)). And seems we don't this issue in the old version🤔. So I want to know whether this statement is correct in the v3.50 version or not.

Thanks a lot. And if needed maybe I can make a pull request to add some locking when reauthenticate.

**Comment by YaoC** [comment]
#13262 fixed the bug but still out of release.

**Comment by zeevogels** [comment]
> #13262 fixed the bug but still out of release.

Why was it not included in v3.5.1? Could it be included in v3.5.2?

**Comment by YaoC** [comment]
> > #13262 fixed the bug but still out of release.
> 
> Why was it not included in v3.5.1? Could it be included in v3.5.2?

This fix was included in [v3.5.2](https://github.com/etcd-io/etcd/blob/v3.5.2/client/v3/retry_interceptor.go#L153:18).
